File: Algorithms/DivideConquer/NumInversions.py

def _merge(nums, left_index, middle_index, right_index):

    inversions = 0

    left_size = middle_index - left_index + 1
    right_size = right_index - middle_index

    left_elements = nums[left_index:left_index + left_size]
    right_elements = nums[middle_index + 1:middle_index + right_size + 1]

    left, right, index = 0, 0, left_index

    while left < left_size and right < right_size:

        if left_elements[left] <= right_elements[right]:

            nums[index] = left_elements[left]
            left += 1

        else:

            nums[index] = right_elements[right]
            right += 1

            inversions += (left_size - left)

        index += 1

    while left < left_size:

        nums[index] = left_elements[left]

        index += 1; left += 1

    while right < right_size:

        nums[index] = right_elements[right]

        index += 1; right += 1

    return inversions

def get_num_inversions(nums, left_index=None, right_index=None):

    count = 0

    if left_index == None or right_index == None:

        return get_num_inversions(nums, 0, len(nums) - 1)

    elif left_index < right_index:

        middle_index = (left_index + right_index) // 2

        count += get_num_inversions(nums, left_index, middle_index)
        count += get_num_inversions(nums, middle_index + 1, right_index)

        count += _merge(nums, left_index, middle_index, right_index)

    return count

def main():

    nums = [3, 15, 61, 11, 7, 9, 2]

    print(get_num_inversions(nums))

if __name__ == '__main__':
    main()

File: Algorithms/DivideConquer/NumInversions.java

import java.util.Arrays;

public class NumInversions {

    private static int merge(int[] nums, int leftIndex, int middleIndex, int rightIndex) {

        int inversions = 0;

        int leftSize = middleIndex - leftIndex + 1;
        int rightSize = rightIndex - middleIndex;

        int leftElements[] = Arrays.copyOfRange(nums, leftIndex, leftIndex + leftSize);
        int rightElements[] = Arrays.copyOfRange(nums, middleIndex + 1, middleIndex + rightSize + 1);

        int left = 0, right = 0, index = leftIndex;

        while (left < leftSize && right < rightSize) {

            if (leftElements[left] <= rightElements[right]) {

                nums[index++] = leftElements[left++];

            } else {

                nums[index++] = rightElements[right++];

                inversions += (leftSize - left);

            }

        }

        while (left < leftSize) {

            nums[index++] = leftElements[left++];

        }

        while (right < rightSize) {

            nums[index++] = rightElements[right++];

        }

        return inversions;

    }

    public static int getNumInversions(int[] nums, int leftIndex, int rightIndex) {

        int count = 0;

        if (leftIndex < rightIndex) {

            int middleIndex = (leftIndex + rightIndex) / 2;

            count += getNumInversions(nums, leftIndex, middleIndex);
            count += getNumInversions(nums, middleIndex + 1, rightIndex);

            count += merge(nums, leftIndex, middleIndex, rightIndex);

        }

        return count;

    }

    public static void main(String[] args) {

        int[] nums = new int[]{3, 15, 61, 11, 7, 9, 2};

        System.out.println(getNumInversions(nums, 0, nums.length - 1));

    }

}
